app/gen-server/lib/HomeDBManager.ts

```diff
@@ -2386,33 +2376,30 @@ export class HomeDBManager extends EventEmitter {
   // maxInheritedRole set on the workspace. Note that information for all users in the org
   // is given to indicate which users have access to the org but not to this particular workspace.
   public async getWorkspaceAccess(scope: Scope, wsId: number): Promise<QueryResult<PermissionData>> {
-    const wsQuery = this._workspace(scope, wsId, {
-      markPermissions: Permissions.VIEW
-    })
-    // Join the workspace's ACL rules (with 1st level groups/users listed).
-    .leftJoinAndSelect('workspaces.aclRules', 'acl_rules')
-    .leftJoinAndSelect('acl_rules.group', 'workspace_groups')
-    .leftJoinAndSelect('workspace_groups.memberUsers', 'workspace_group_users')
-    .leftJoinAndSelect('workspace_groups.memberGroups', 'workspace_group_groups')
-    .leftJoinAndSelect('workspace_group_users.logins', 'workspace_user_logins')
-    // Join the org and groups/users.
-    .leftJoinAndSelect('workspaces.org', 'org')
-    .leftJoinAndSelect('org.aclRules', 'org_acl_rules')
-    .leftJoinAndSelect('org_acl_rules.group', 'org_groups')
-    .leftJoinAndSelect('org_groups.memberUsers', 'org_group_users')
-    .leftJoinAndSelect('org_group_users.logins', 'org_user_logins');
-    const queryResult = await verifyIsPermitted(wsQuery);
-    if (queryResult.status !== 200) {
-      // If the query for the doc failed, return the failure result.
-      return queryResult;
+    const wsQueryResult = await this._getWorkspaceWithACLRules(scope, wsId);
+    if (wsQueryResult.status !== 200) {
+      // If the query for the workspace failed, return the failure result.
+      return wsQueryResult;
     }
-    const workspace: Workspace = queryResult.data;
+    const workspace: Workspace = wsQueryResult.data;
+
     const wsMap = getMemberUserRoles(workspace, this.defaultCommonGroupNames);
+
+    // Also fetch the organization ACLs so we can determine inherited rights.
+    const orgQuery = this._buildOrgWithACLRulesQuery(scope, workspace.org.id);
+    const orgQueryResult = await verifyEntity(orgQuery, { discardPermissionsCheck: true });
```

Comment:
Do you have any concerns about changes to the database that may have happened between this query and the previous one?

Reply:
To be sure, by "the previous one", do you mean:
 - previous in the code? (which would refer to the workspace ACL selection)
 - previous in the commits? (which would refer to the org ACL selection with permission check)

Reply:
Previous in the code.

Reply:
To be clearer, I'm asking about this sequence:
 * Query 1 happens
 * Orgs/workspaces/documents/users... are created/updated/deleted
 * Query 2 happens

Reply:
Right. I am no DB expert, my guess is that the drawback of this change is the lack of isolation.

To limit that drawback, I guess we might otherwise gather these two queries into one using `UNION`, though I don't know how much work it would represent.

I am scratching a bit my head tomorrow, I can challenge a bit more myself about it, or ask the opinions of colleagues.

Reply:
You could do both queries in a transaction. There's a few examples in this file (e.g. `this._connection.transaction(...)`).

Reply:
I have used a transaction. That being said, I am a bit divided about the isolation level to set. I let the database use its default one (`READ COMMIT` for pg and `SERIALIZABLE` for sqlite).

With the hope it is acceptable regarding the performance, I propose to use the SERIALIZABLE isolation level (which is the default one for sqlite but not for postgresql).

I am still a bit divided, I would say that if there are some resources modified in between, as the function is only used for the API (and client-side by the ACL popup), the user might be slightly misled but not much (they may see outdated rights in workspaces). If the organization is deleted in between, they should see a 404 error. 

Still I prefer being prudent and put a SERIALIZABLE isolation level to be sure…

Reply:
> I am still a bit divided, I would say that if there are some resources modified in between, as the function is only used for the API (and client-side by the ACL popup), the user might be slightly misled but not much (they may see outdated rights in workspaces). If the organization is deleted in between, they should see a 404 error.

The important thing is consistency. Without the transaction, it is hard to reason about what the call could return, because a lot could happen in between the queries. With the transaction, it is easier to think about what it will do. Whether the result is more or less stale is not that important, to my mind, because no matter what you do, by the time the caller receives the reply there's no longer any guarantee it is still valid - anything could have happened in the meantime.

Reply:
Thanks for your feedback. I am sorry, I am not sure to understand your remark, is it OK what I did or should I bring some change to my PR?

Reply:
I think it looks ok @fflorent. Agree with @paulfitz that the isolation level isn't that important here, so ok to go with the default. Thanks.

Reply:
Done! [690fdce](https://github.com/gristlabs/grist-core/pull/824/commits/690fdcec86cb3a45027db3c2bd5a405de3536407)

Reply:
Thanks @fflorent !